openvdb/openvdb/tree/LeafManager.h

```diff
@@ -635,34 +615,99 @@ class LeafManager
 
 private:
 
-    // This a simple wrapper for a c-style array so it mimics the api
-    // of a std container, e.g. std::vector or std::deque, and can be
-    // passed to Tree::getNodes().
-    struct MyArray {
-        using value_type = LeafType*;//required by Tree::getNodes
-        value_type* ptr;
-        MyArray(value_type* array) : ptr(array) {}
-        void push_back(value_type leaf) { *ptr++ = leaf; }//required by Tree::getNodes
-    };
-
-    void initLeafArray()
+    void initLeafArray(bool serial = false)
     {
-        const size_t leafCount = mTree->leafCount();
+        // Build an array of all nodes that have leaf nodes as their immediate children
+
+        using NodeChainT = typename NodeChain<RootNodeType, RootNodeType::LEVEL>::Type;
+        using NonConstLeafParentT = typename NodeChainT::template Get</*Level=*/1>;
+        using LeafParentT = typename CopyConstness<TreeType, NonConstLeafParentT>::Type;
+
+        std::deque<LeafParentT*> leafParents;
+        mTree->getNodes(leafParents);
+
+        // Compute the leaf counts for each node
+
+        std::vector<Index32> leafCounts;
+        if (serial) {
+            leafCounts.reserve(leafParents.size());
+            for (LeafParentT* leafParent : leafParents) {
+                leafCounts.push_back(leafParent->childCount());
+            }
+        } else {
+            leafCounts.resize(leafParents.size());
+            tbb::parallel_for(
+                tbb::blocked_range<size_t>(0, leafParents.size(), /*grainsize=*/64),
```

Comment:
remove arbitrary grain size maybe? Or expose as option

Reply:
I tend to follow these two bits of advice from TBB:

* A rule of thumb is that grainsize iterations of operator() should take at least 100,000 clock cycles to execute
* You do not have to set the grainsize too precisely

It's not arbitrary - my general approach is to pick between grain sizes of 1, 64 and 1024 and err on the low side. In this case, I opted for 64 as for a typical node size, this will translate to 64 popcount() instructions which will likely not be sufficient to maximise throughput. You'll notice that the subsequent parallel_for has no grain size because there is plenty of work being done by the operation, hence why exposing it doesn't really make sense.

Reply:
I would then add a comment as to why 64 is a better value here from your reasons given, otherwise it does look arbitrary and could very well have come from a value you chose based on your particular hardware.

Reply:
Sure, I can add a quick comment about this here.

Reply:
Thanks - I think as a rule, if you select a particular grainsize other than 1, then a brief comment should accompany it